Fix possible typo with `gnupg-curl` vs `gnupg curl`

zahiar opened the pull request:
### What does this PR do?
Fixes an issue with salt not installing on Ubuntu 16, failing on `install_ubuntu_stable_deps`.  The issue lay with a possible suspect typo, where it was installing just `gnupg-curl`, but looking at the lines above, that was installing two packages `gnupg dirmngr`.

So installing `curl` fixed the issue, as it looks to be reliant upon a sub-dependency of it.

### What issues does this PR fix or reference?

### Previous Behavior
`install_ubuntu_stable_deps` exits unsuccessfully..

```
Setting up ca-certificates (20170717~16.04.2) ...
Setting up krb5-locales (1.13.2+dfsg-5ubuntu2.1) ...
Setting up libsasl2-modules:amd64 (2.1.26.dfsg1-14ubuntu0.1) ...
Setting up gnupg-curl (1.4.20-1ubuntu3.3) ...
Processing triggers for libc-bin (2.23-0ubuntu11) ...
Processing triggers for ca-certificates (20170717~16.04.2) ...
Updating certificates in /etc/ssl/certs...
148 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
 * ERROR: Failed to run install_ubuntu_stable_deps()!!!
 * DEBUG: Removing the logging pipe /tmp/bootstrap-salt.logpipe
 * DEBUG: Removing the temporary apt error file /tmp/apt_error.n3jI
ERROR: Service 'ubuntu-16-tester' failed to build: The command '/bin/sh -c /tmp/bootstrap-salt.sh -X -d -D' returned a non-zero code: 1
```

### New Behavior
`install_ubuntu_stable_deps` now runs successfully.

```
Setting up ca-certificates (20170717~16.04.2) ...
Setting up krb5-locales (1.13.2+dfsg-5ubuntu2.1) ...
Setting up libsasl2-modules:amd64 (2.1.26.dfsg1-14ubuntu0.1) ...
Setting up curl (7.47.0-1ubuntu2.13) ...
Processing triggers for libc-bin (2.23-0ubuntu11) ...
Processing triggers for ca-certificates (20170717~16.04.2) ...
Updating certificates in /etc/ssl/certs...
148 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
OK
```

zahiar committed:
Fix possible typo with `gnupg-curl` vs `gnupg curl`

Installing just `gnupg-curl` causes this issue:
```
Setting up ca-certificates (20170717~16.04.2) ...
Setting up krb5-locales (1.13.2+dfsg-5ubuntu2.1) ...
Setting up libsasl2-modules:amd64 (2.1.26.dfsg1-14ubuntu0.1) ...
Setting up gnupg-curl (1.4.20-1ubuntu3.3) ...
Processing triggers for libc-bin (2.23-0ubuntu11) ...
Processing triggers for ca-certificates (20170717~16.04.2) ...
Updating certificates in /etc/ssl/certs...
148 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
 * ERROR: Failed to run install_ubuntu_stable_deps()!!!
 * DEBUG: Removing the logging pipe /tmp/bootstrap-salt.logpipe
 * DEBUG: Removing the temporary apt error file /tmp/apt_error.n3jI
ERROR: Service 'ubuntu-16-tester' failed to build: The command '/bin/sh -c /tmp/bootstrap-salt.sh -X -d -D' returned a non-zero code: 1
```

Where as installing both `gnupg` and `curl` ensures `install_ubuntu_stable_deps` runs correctly, as a sub-dependency of `curl` looks to be missing.
```
Setting up ca-certificates (20170717~16.04.2) ...
Setting up krb5-locales (1.13.2+dfsg-5ubuntu2.1) ...
Setting up libsasl2-modules:amd64 (2.1.26.dfsg1-14ubuntu0.1) ...
Setting up curl (7.47.0-1ubuntu2.13) ...
Processing triggers for libc-bin (2.23-0ubuntu11) ...
Processing triggers for ca-certificates (20170717~16.04.2) ...
Updating certificates in /etc/ssl/certs...
148 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
OK
```

I suspect this is a typo, because looking at the code line above, `__PACKAGES="${__PACKAGES} gnupg dirmngr"`, that installs two packages.

nielssorensen commented:
gnupg-curl is a package but it does not seem to be one required for this.  By removing the hyphen, you are telling apt(-get) to install gnupg AND curl.  If you remove gnupg and leave only curl, this also works.

The entire block was added as part of this commit, 1083a5c17fcb0d0a200a3bc77de7f4a64c0b79e7, for "Fix #1137: import GPG key through an HTTP(S) proxy".

s0undt3ch commented:
/cc @vutny

s0undt3ch commented:
Sorry, shouldn't just CC'd, could you please take a look and clarify @vutny?

vutny reviewed (changes requested):
Hi @zahiar 

Thanks for your PR.
The code installing `dirmngr` and `gnupg-curl` is simply obsoleted, these tools are no longer in use after the fix to APT GPG key importing function was merged.

But instead, the script relies on `wget` or `curl` being already installed before bootstrapping on Ubuntu.
Since `wget` is a standard package in Ubuntu Server, everything works in majority of use cases.
I assume you use minimal installation, some image or Docker, so no tools for downloading the key are available.

I suggest just to attempt to install `wget` on all Ubuntu version. So if it is already installed, the system would not be populated with probably unwanted packages. The `gnupg` is a package required by distribution, no need to explicitly add it.

zahiar committed:
Update downloader to just install `wget`

This is the only package required to down they keys and also, there's no need to include `gnupg` package explicitly as its required by the distribution anyway.

zahiar commented:
@vutny Thanks for your feedback!

Installing `wget` across all Ubuntu versions works wonderfully well, even on the Docker setup I have 😄 

However, it looks as though we will have to install `gnupg` on Debian 9+ & Ubuntu 18.04+ 😞 

Debian 9 CI log
```Setting up apt-transport-https (1.4.9) ...
309 | E: gnupg, gnupg2 and gnupg1 do not seem to be installed, but one of them is required for this operation
310 | * ERROR: Failed to run install_debian_9_git_deps()!!!
311 | No salt-minion installed, install must have failed!!```

Ubuntu 18.04 CI log
```Setting up apt-transport-https (1.6.11) ...
319 | E: gnupg, gnupg2 and gnupg1 do not seem to be installed, but one of them is required for this operation
320 | * ERROR: Failed to run install_ubuntu_git_deps()!!!```

So I'll add them back in

zahiar committed:
Install required `gnupg` package for later Debian/Ubuntu distros

Debian 9+ & Ubuntu 18+ do not appear to install an `gnupg` or equivalent package by default, therefore we need to install it instead, as it is required for importing the GPG keys.

vutny reviewed (approved):
Awesome, @zahiar .
Thank you for spotting it on Debian as well.

zahiar commented:
Eek, looks like the CI job has been running for over 26 hours and has become stuck 🙈 
I suspect it's docker related by looking at the CI job logs 🤔 

In any case, anyone know how to cancel the current job (cancel button is disabled for me) and re-run it?

zahiar commented:
Thanks to whoever re-ran the CI job - all checks have passed 👍

zahiar commented:
@s0undt3ch @vutny as this PR has been approved and has passed all checks - may we get an update on getting it merged in please? 😊 

Thanks in advance 😄
